### PR TITLE
chore: login per keycloak via OAuth2ClientBundle

### PR DESCRIPTION
**Pull request by mussbach**
### Ticket
DPLAN-11848

This PR moves the authentication from login via SAML per OneLoginSamlBundle to login per OAuth via OAuth2ClientBundle.

It is mostly a c&p from `SamlUserFactory`, `SamlUserProvider` and friends to new classes that use the OAuth2ClientBundle. Alon the way a lot of tests where added.
It uses the same basis of the value objects like the `OzgKeycloakAuthenticator`, but keeps the logic from the OneLoginSamlBundle whenever possible.

Later on it might be a good idea to merge `OzgKeycloakAuthenticator` and `KeycloakAuthenticator` (and subsequent classes), but to keep this PR as small as possible this was not done yet.

### How to review/test
code review or install the branch at the release server and test login

### PR Checklist

Delete the checkbox if it doesn't apply/isn't necessary.

- [x] Tests updated/created
- [x] Update documentation
- [x] Link all relevant tickets
- [x] Move the tickets on the board accordingly
- [ ] Data-Cy attributes added/updated ([conventions](https://dplan-documentation.demos-europe.eu/development/guidelines-conventions/coding-styleguides/twig_html.html#guideline-for-naming-cypress-hooks))

### CODE REVIEW

**Review by MoritzMandler — DISMISSED** [review]
lgtm

File: demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/KeycloakAuthenticator.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        // propagate user login to session
+        $request->getSession()->set('userId', $user->getId());
+
+        // change "app_homepage" to some route in your app
+        $targetUrl = $this->router->generate('core_home_loggedin');
+
+        return new RedirectResponse($targetUrl);
+
+        // or, on success, let the request continue to be handled by the controller
+        // return null;
```
Comment (breierDemos) [inline demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/KeycloakAuthenticator.php:74]: commented out code

File: demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/OzgKeycloakAuthenticator.php [hunk 2]
```diff
@@ -35,7 +35,7 @@ class OzgKeycloakAuthenticator extends OAuth2Authenticator implements Authentica
     public function __construct(
         private readonly ClientRegistry $clientRegistry,
```
Comment (breierDemos) [inline demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/OzgKeycloakAuthenticator.php:36]: The "p" in AuthenticationEntrypointInterface should be uppercase to match declaration.

File: demosplan/DemosPlanCoreBundle/ValueObject/KeycloakUserData.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
+<?php
```
Comment (breierDemos) [inline demosplan/DemosPlanCoreBundle/ValueObject/KeycloakUserData.php:1]: Strict types declaration is missing

File: tests/backend/core/Core/Functional/KeycloakUserDataTest.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            $this->keycloakUserData->getOrganisationName()
+        );
+        $this->assertEquals('456', $this->keycloakUserData->getUserId());
+        $this->assertEquals(self::TEST_EMAIL, $this->keycloakUserData->getUserName());
+        $this->assertEquals('10', $this->keycloakUserData->getHouseNumber());
+        $this->assertEquals('Test City', $this->keycloakUserData->getLocalityName());
+        $this->assertEquals('12345', $this->keycloakUserData->getPostalCode());
+        $this->assertEquals('Test Street', $this->keycloakUserData->getStreet());
+
+    }
```
Comment (breierDemos) [inline tests/backend/core/Core/Functional/KeycloakUserDataTest.php:62]: Empty Line before closing bracket.

File: tests/backend/core/Core/Functional/KeycloakUserDataTest.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    public function testUserInformationIsCorrectlyFilledFromIncompleteResourceOwner(): void
```
Comment (breierDemos) [inline tests/backend/core/Core/Functional/KeycloakUserDataTest.php:64]: test failed

File: tests/backend/core/Core/Functional/KeycloakUserDataMapperTest.php [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+use demosplan\DemosPlanCoreBundle\Entity\User\AnonymousUser;
+use demosplan\DemosPlanCoreBundle\Entity\User\User;
+use demosplan\DemosPlanCoreBundle\EventDispatcher\EventDispatcherPostInterface;
+use demosplan\DemosPlanCoreBundle\Logic\KeycloakUserDataMapper;
+use demosplan\DemosPlanCoreBundle\Logic\User\CustomerService;
+use demosplan\DemosPlanCoreBundle\Logic\User\OrgaService;
+use demosplan\DemosPlanCoreBundle\Logic\User\RoleHandler;
+use demosplan\DemosPlanCoreBundle\Logic\User\UserService;
+use demosplan\DemosPlanCoreBundle\ValueObject\KeycloakUserData;
+use Liip\FunctionalTestBundle\Test\WebTestCase;
```
Comment (breierDemos) [inline tests/backend/core/Core/Functional/KeycloakUserDataMapperTest.php:18]: WebTestCase not used

File: tests/backend/core/Core/Functional/KeycloakUserDataMapperTest.php [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    private ?KeycloakUserDataMapper $keycloakUserDataMapper;
+
+    /**
+     * @var User
+     */
+    protected $citizenUser;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
```
Comment (breierDemos) [inline tests/backend/core/Core/Functional/KeycloakUserDataMapperTest.php:36]: setup failed here

**Review by breierDemos — CHANGES_REQUESTED** [review]
Code Review Only

File: demosplan/DemosPlanCoreBundle/Logic/KeycloakUserDataMapper.php [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+        $user = $this->userService->findDistinctUserByEmailOrLogin($keycloakUserData->getEmailAddress());
+
+        if ($user instanceof User) {
+            $this->logger->info('Found user in Keycloak request', ['data' => $keycloakUserData]);
+
+            $this->updateOrgaWithKnownValues($user->getOrga(), $keycloakUserData);
+
+            return $this->updateUserWithKnownValues($user, $keycloakUserData);
+
+        }
```
[the remark this answers is not in the document]
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Logic/KeycloakUserDataMapper.php:65]: f67bf45

File: demosplan/DemosPlanCoreBundle/Logic/KeycloakUserDataMapper.php [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+        if ('' !== $keycloakUserData->getOrganisationName()) {
+            $this->logger->info('User is Orgauser');
+            return $this->getUserForOrga($keycloakUserData);
+        }
+
+        // otherwise we assume that it is a citizen
+        $this->logger->info('User is citizen');
+        return $this->getUserForNewCitizen($keycloakUserData);
+
+    }
```
[the remark this answers is not in the document]
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Logic/KeycloakUserDataMapper.php:85]: f67bf45

File: demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/OzgKeycloakAuthenticator.php (hunk 2 above)
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/OzgKeycloakAuthenticator.php:36]: f67bf45

File: tests/backend/core/Core/Functional/KeycloakUserDataMapperTest.php (hunk 7 above)
Comment (mussbach) [reply tests/backend/core/Core/Functional/KeycloakUserDataMapperTest.php:43]: works for me, we need to check your setup. What fails?

File: tests/backend/core/Core/Functional/KeycloakUserDataTest.php (hunk 4 above)
Comment (mussbach) [reply tests/backend/core/Core/Functional/KeycloakUserDataTest.php:69]: f67bf45

File: tests/backend/core/Core/Functional/KeycloakUserDataTest.php (hunk 5 above)
Comment (mussbach) [reply tests/backend/core/Core/Functional/KeycloakUserDataTest.php:71]: What is the reason? It does not for me

File: demosplan/DemosPlanCoreBundle/ValueObject/KeycloakUserData.php (hunk 3 above)
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/ValueObject/KeycloakUserData.php:1]: [b8801fa](https://github.com/demos-europe/demosplan-core/pull/3284/commits/b8801fa7e86c2fe396cfe65b0c47c2ae993820e7)

File: demosplan/DemosPlanCoreBundle/Logic/KeycloakUserDataMapper.php [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+        $userEmail = $keycloakUserData->getEmailAddress();
+        $orgaTypeNames = [OrgaType::PUBLIC_AGENCY];
+
+        $orga = $this->orgaService->createOrgaRegister(
+            $orgaName,
+            $phone,
+            $userFirstName,
+            $userLastName,
+            $userEmail,
+            $this->customerService->getCurrentCustomer(),
```
[the remark this answers is not in the document]
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Logic/KeycloakUserDataMapper.php]: [b8801fa](https://github.com/demos-europe/demosplan-core/pull/3284/commits/b8801fa7e86c2fe396cfe65b0c47c2ae993820e7)

File: demosplan/DemosPlanCoreBundle/Logic/KeycloakUserDataMapper.php [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+use demosplan\DemosPlanCoreBundle\Entity\User\User;
+use demosplan\DemosPlanCoreBundle\Event\User\NewOrgaRegisteredEvent;
+use demosplan\DemosPlanCoreBundle\EventDispatcher\EventDispatcherPostInterface;
+use demosplan\DemosPlanCoreBundle\Exception\InvalidArgumentException;
+use demosplan\DemosPlanCoreBundle\Logic\User\CustomerService;
+use demosplan\DemosPlanCoreBundle\Logic\User\OrgaService;
+use demosplan\DemosPlanCoreBundle\Logic\User\RoleHandler;
+use demosplan\DemosPlanCoreBundle\Logic\User\UserService;
+use demosplan\DemosPlanCoreBundle\ValueObject\KeycloakUserData;
+use demosplan\DemosPlanCoreBundle\ValueObject\KeycloakUserDataInterface;
```
[the remark this answers is not in the document]
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Logic/KeycloakUserDataMapper.php]: [f67bf45](https://github.com/demos-europe/demosplan-core/pull/3284/commits/f67bf451718ecef4d26bf27e4a50b55082973cbc)

File: demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/KeycloakAuthenticator.php [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+use KnpU\OAuth2ClientBundle\Client\ClientRegistry;
+use KnpU\OAuth2ClientBundle\Security\Authenticator\OAuth2Authenticator;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
```
[the remark this answers is not in the document]
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/KeycloakAuthenticator.php]: [f67bf45](https://github.com/demos-europe/demosplan-core/pull/3284/commits/f67bf451718ecef4d26bf27e4a50b55082973cbc)

File: demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/KeycloakAuthenticator.php [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+use Symfony\Component\Security\Http\Authenticator\Passport\Passport;
+use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;
+use Symfony\Component\Security\Http\EntryPoint\AuthenticationEntryPointInterface;
+
+class KeycloakAuthenticator extends OAuth2Authenticator implements AuthenticationEntrypointInterface
```
[the remark this answers is not in the document]
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/KeycloakAuthenticator.php]: [f67bf45](https://github.com/demos-europe/demosplan-core/pull/3284/commits/f67bf451718ecef4d26bf27e4a50b55082973cbc)

File: demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/KeycloakAuthenticator.php (hunk 1 above)
Comment (mussbach) [reply demosplan/DemosPlanCoreBundle/Security/Authentication/Authenticator/KeycloakAuthenticator.php]: [f67bf45](https://github.com/demos-europe/demosplan-core/pull/3284/commits/f67bf451718ecef4d26bf27e4a50b55082973cbc)

File: tests/backend/core/Core/Functional/KeycloakUserDataTest.php (hunk 5 above)
Comment (breierDemos) [reply tests/backend/core/Core/Functional/KeycloakUserDataTest.php:71]: Okay if it works, then I have to check my setup.

File: tests/backend/core/Core/Functional/KeycloakUserDataMapperTest.php (hunk 7 above)
Comment (breierDemos) [reply tests/backend/core/Core/Functional/KeycloakUserDataMapperTest.php:43]: Okay if it works, then I have to check my setup.

**Review by breierDemos — APPROVED** [review]
lgtm I will check my test setup
remaining concerns can be handled with another pr

**Pull request merged**